presto-main/src/main/java/com/facebook/presto/metadata/Metadata.java

```diff
@@ -157,6 +157,12 @@
      */
     void createTable(Session session, String catalogName, ConnectorTableMetadata tableMetadata, boolean ignoreExisting);
 
+    /**
+     * Creates a temporary table with optionally partitioning requirements can be specified.
+     * Temporary table gets automatically removed when the current transaction ends.
+     */
+    TableHandle createTemporaryTable(Session session, String catalogName, List<ColumnMetadata> columns, Optional<PartitioningMetadata> partitioning);
```

Comment:
I have some thoughts about this new introduced API.

A developer not familiar with out work might come up with this question: : Why the interface looks different with `createTable`, or even, can we merge these two interfaces? (by a boolean flag). Since IIRC, the implementation in `HiveMetadata` is also quite familiar. 

If we look into `ConnectorTableMetadata` class (used by `createTable`):

```java
public class ConnectorTableMetadata
{
    private final SchemaTableName table;
    private final Optional<String> comment;
    private final List<ColumnMetadata> columns;
    private final Map<String, Object> properties;
}
```

`SchemaTableName` is not a big deal -- we can make engine come up with the random name, ask Connector to provide a temporary table name, or even just make it Optional (not preferred).

`comment` is already optional.

`columns` is also provided in the `createTemporaryTable` API as well.

The real problem is, today table partitioning information is not a first-citizen class, and only described in in `properties` in a connector-specific way. In Hive, its name is `bucketed_count` and `bucketed_by`. 

In this commit you already introduced `ConnectorPartitioningMetadata` (which, is kind of the `bucketed_count` and `bucketed_by`). Now this opens opportunity to just make `ConnectorPartitioningMetadata` an optional  field in `ConnectorTableMetadata`. So table partitioning information is no longer an opaque thing encoded in the `properties` map  -- and we can unify the `createTable` and `createTemporaryTable` API.

What about the following:
- We still introduce this not-so-elegant API, but mark it as unstable and warn people not to use it. Since this refactor (to use `ConnectorPartitioningMetadata` in `ConnectorTableMetadata`) might require some time and some corner cases, and sounds like a separate project. I would also like to view this as an opportunity to help other engineers ramp up :) 
- Once this is merged, @rschlussel can help refactor this API.

It might also be helpful to discuss in person with other committers next week.  cc @highker, @rongrong, @shixuan

Reply:
@wenleix I agree with marking the new SPI (`createTemporaryTable`) as experimental. However I'm less sure about merging both `createTable` and `createTemporaryTable` into a single method. Temporary tables have different semantics (temporary tables live only for a duration of a session). So in every connector that implements `createTable`, but doesn't support temporary table concept we will be adding a `check(!metadata.isTemporary())` check , to avoid confusions.

Reply:
Why not just use `createTable` directly? Or `createTemporaryTable` is a syntactic sugar for `createTable` + `dropTable`?

Reply:
@highker That's a very good question.

There many reasons. Here are some of them sorted by importance:

1. `createTable` doesn't accept partitioning handle. Instead it accepts connector specific properties. Different connector may have different properties that are used to declare some sort of analogue of "bucketing"
2. Relying on Transaction manager to cleanup temporary tables simplifies implementation. otherwise we would have to implement yet another post-processing. Maybe in the future we will think of decreasing the lifespan of a temporary table (by explicitly calling "delete" after we consider a table as no longer needed).
3. Temporary table shoudn't be an "opaque" thing for a connector. Connector may choose to use different file format, replication level, etc. for the temporary table. Simply adding a flag to the "TableMetadata" seems to be not a very good idea. As other connectors may not support temporary tables, and never check that flag.
4. It should be up to a connector to assign the temporary table name and the temporary table schema. It will be cumbersome to assign this values by the engine. As it is harder to check if the chosen name not yet exist.
5. "TableMetadata" accepts a list of connector specific properties, that are meaningless in a temporary table case.

Reply:
> However I'm less sure about merging both createTable and createTemporaryTable into a single method. 

Sure. That's more debatable  ;) . What do you think about my proposal to make `ConnectorPartitioningMetadata` part of `ConnectorTableMetadata` ? ;)

Reply:
> 1. createTable doesn't accept partitioning handle. Instead it accepts connector specific properties. Different connector may have different properties that are used to declare some sort of analogue of "bucketing"

This can be solved by my "putting ConnectorPartitioningMetadata` into  ConnectorTableMetadata" proposal right? ;) 

> 3. Temporary table shoudn't be an "opaque" thing for a connector. Connector may choose to use different file format, replication level, etc. for the temporary table. Simply adding a flag to the "TableMetadata" seems to be not a very good idea. As other connectors may not support temporary tables, and never check that flag.
> 5. "TableMetadata" accepts a list of connector specific properties, that are meaningless in a temporary table case.

I agree some (or most) properties doesn't  make sense to Temporary Table.... hmm... So maybe it deserves a separate API :) 

--------------------